### PR TITLE
Rewrite CosmosVFS

### PR DESCRIPTION
**Pull request by MishaTY**
Rewrites CosmosVFS class with a new API. You can easily get all of the disks and partitions inside of those disks.

TODO:

-  [x] Add better partition support
- [x] Add partition to disk
- [x] Remove partition
- [x] Add remove/add partitions functions to Disk.cs

### CODE REVIEW

File: source/Cosmos.HAL2/BlockDevice/Partition.cs [hunk 1]
```diff
@@ -10,7 +10,7 @@ public class Partition : BlockDevice
         /// <summary>
         /// Hosting device.
         /// </summary>
-        private readonly BlockDevice mHost;
+        public readonly BlockDevice mHost;
```
Comment (quajak) [inline source/Cosmos.HAL2/BlockDevice/Partition.cs:13]: Please rename this to `Host`

File: source/Cosmos.System2/FileSystem/CosmosVFS.cs [hunk 2]
```diff
@@ -38,29 +39,32 @@ public class CosmosVFS : VFSBase
...
         public override void Initialize()
         {
-            mPartitions = new List<Partition>();
-            mFileSystems = new List<FileSystem>();
-            mRegisteredFileSystems = new List<FileSystemFactory>();
-
-            RegisterFileSystem(new FatFileSystemFactory());
+            if (BlockDevice.Devices.Count == 0)
+            {
+                global::System.Console.WriteLine("No disks found!");
```
Comment (quajak) [inline source/Cosmos.System2/FileSystem/CosmosVFS.cs:44]: This should be suppressible, in which case an exception is thrown for the user os to handle.

File: source/Cosmos.System2/FileSystem/CosmosVFS.cs [hunk 3]
```diff
@@ -821,7 +737,11 @@ public override bool IsValidDriveId(string driveId)
         /// <exception cref="Exception">Unable to determine filesystem for path:  + aDriveId</exception>
         public override long GetTotalSize(string aDriveId)
         {
-            var xFs = GetFileSystemFromPath(aDriveId);
+            var xFs = GetPartitionFromPath(aDriveId).MountedFS;
```
Comment (quajak) [inline source/Cosmos.System2/FileSystem/CosmosVFS.cs:740]: Can you please refractor the getting and checking of the filesystem into a helper function to reduce repeated code?

File: source/Cosmos.System2/FileSystem/CosmosVFS.cs [hunk 4]
```diff
@@ -836,7 +756,11 @@ public override long GetTotalSize(string aDriveId)
         /// <exception cref="Exception">Unable to determine filesystem for path:  + aDriveId</exception>
         public override long GetAvailableFreeSpace(string aDriveId)
```
Comment (quajak) [inline source/Cosmos.System2/FileSystem/CosmosVFS.cs:757]: Do we need these wrapper functions? Or it enough if you can get the file system from a drive id to query the data yourself?

File: source/Cosmos.System2/FileSystem/Disk.cs [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+        /// <summary>
+        /// Get the volume separator char.
+        /// </summary>
+        public static char VolumeSeparatorChar { get { return ':'; } }
+        public static List<FileSystemFactory> registeredFileSystems = new List<FileSystemFactory>();
+        /// <summary>
+        /// List of file systems.
+        /// </summary>
+        public static List<FileSystemFactory> RegisteredFileSystems { get { return registeredFileSystems; } }
```
Comment (quajak) [inline source/Cosmos.System2/FileSystem/Disk.cs:39]: Please rename to `RegisteredFileSystemsTypes`

File: source/Cosmos.System2/FileSystem/ManagedPartition.cs [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+
+        public ManagedPartition(Partition host)
+        {
+            Host = host;
+        }
+        /// <summary>
+        /// Mounts the partition
+        /// </summary>
+        public void Mount()
```
Comment (quajak) [inline source/Cosmos.System2/FileSystem/ManagedPartition.cs:45]: Maybe also add a version of the Mount function where you pass the FilesystemFactory in as well

**Review by quajak — CHANGES_REQUESTED** [review]
Thanks for the PR! The idea is solid, just a bit of tweaking/polishing left. Could you please also add some tests to the filesystem testrunner?

File: source/Cosmos.System2/FileSystem/Disk.cs [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    public class Disk
+    {
+        /// <summary>
+        /// The size of the disk in MB.
+        /// </summary>
+        public int Size { get; }
+        /// <summary>
+        /// List of partitions
+        /// </summary>
+        public List<ManagedPartition> parts = new List<ManagedPartition>();
```
[the remark this answers is not in the document]
Comment (MishaTY) [reply source/Cosmos.System2/FileSystem/Disk.cs:17]: Oops, I forgot to make that private.

File: source/Cosmos.System2/FileSystem/Disk.cs [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+
+        /// <summary>
+        /// Get the directory separator char.
+        /// </summary>
+        public static char DirectorySeparatorChar { get { return '\\'; } }
+
+        /// <summary>
+        /// Get the alt. directory separator char.
+        /// </summary>
+        public static char AltDirectorySeparatorChar { get { return '/'; } }
```
[the remark this answers is not in the document]
Comment (MishaTY) [reply source/Cosmos.System2/FileSystem/Disk.cs:29]: No, we don't. I was just copying and pasting code from the old CosmosVFS and forgot to remove that.

File: source/Cosmos.System2/FileSystem/Disk.cs [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+        /// List of partitions
+        /// </summary>
+        public List<ManagedPartition> parts = new List<ManagedPartition>();
+
+        public List<ManagedPartition> Partitions { get { return parts; } }
+
+        /// <summary>
+        /// Get the directory separator char.
+        /// </summary>
+        public static char DirectorySeparatorChar { get { return '\\'; } }
```
[the remark this answers is not in the document]
Comment (MishaTY) [reply source/Cosmos.System2/FileSystem/Disk.cs:24]: Ok, I will move that.

File: source/Cosmos.System2/FileSystem/CosmosVFS.cs (hunk 4 above)
Comment (MishaTY) [reply source/Cosmos.System2/FileSystem/CosmosVFS.cs:751]: I think it's better to completely remove the VFSBase and CosmosVFS classes and replace them with a DiskManager class where the user can get all of the disks.

**Comment by MishaTY** [comment]
This PR is ready for review and merging now.

File: source/Cosmos.HAL2/BlockDevice/IDE.cs [hunk 10]
```diff
@@ -40,62 +40,63 @@ private static void Initialize(Ata.ControllerIdEnum aControllerID, Ata.BusPositi
...
             {
-                var xGPT = new GPT(xATA);
+                var xGPT = new GPT(device);
 
                 Ata.AtaDebugger.Send("Number of GPT partitions found:");
                 Ata.AtaDebugger.SendNumber(xGPT.Partitions.Count);
-                for (int i = 0; i < xGPT.Partitions.Count; i++)
+                var parts = xGPT.GetPartitions();
+                for (int i = 0; i < parts.Count; i++)
                 {
```
Comment (quajak) [inline source/Cosmos.HAL2/BlockDevice/IDE.cs:56]: Can this be a for loop?

File: source/Cosmos.HAL2/BlockDevice/IDE.cs [hunk 11]
```diff
@@ -40,62 +40,63 @@ private static void Initialize(Ata.ControllerIdEnum aControllerID, Ata.BusPositi
...
-                    var xPart = xGPT.Partitions[i];
+                    var xPart = parts[i];
                     if (xPart == null)
```
Comment (quajak) [inline source/Cosmos.HAL2/BlockDevice/IDE.cs:58]: Also does the current `GetPartitions` allow for null partitions to happen?

File: source/Cosmos.HAL2/BlockDevice/IDE.cs [hunk 12]
```diff
@@ -40,62 +40,63 @@ private static void Initialize(Ata.ControllerIdEnum aControllerID, Ata.BusPositi
...
                     }
                 }
-
-                // TODO Change this to foreach when foreach is supported
+                var partitions = mbr.GetPartitions();
                 Ata.AtaDebugger.Send("Number of MBR partitions found:");
-                Ata.AtaDebugger.SendNumber(xMBR.Partitions.Count);
-                for (int i = 0; i < xMBR.Partitions.Count; i++)
+                Ata.AtaDebugger.SendNumber(partitions.Count);
+                for (int i = 0; i < partitions.Count; i++)
```
Comment (quajak) [inline source/Cosmos.HAL2/BlockDevice/IDE.cs:90]: Same question here

**Review by quajak — APPROVED** [review]
Looks good. I think charles should take a look as well before we merge.

**Comment by charlesbetros** [comment]
I think we should keep the VFS base class and use it everywhere. This way users can implement their own VFS if they want.

**Comment by MishaTY** [comment]
I didn't remove the VFSBase class. Users can also implement custom drive letters now.

File: source/Cosmos.System2/FileSystem/Disk.cs [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+        /// </summary>
+        public static List<FileSystemFactory> RegisteredFileSystemsTypes { get { return registeredFileSystems; } }
+        /// <summary>
+        /// Main blockdevice that has all of the partitions.
+        /// </summary>
+        public BlockDevice Host;
+        /// <summary>
+        /// The Master Boot record (MBR) of the drive.
+        /// </summary>
+        public byte[] MBR
```
Comment (quajak) [inline source/Cosmos.System2/FileSystem/Disk.cs:36]: Why is this required? Isnt accessing the `PartitioningType` enough?

File: source/Cosmos.System2/FileSystem/Disk.cs [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+
+        public PartitioningType PartitioningType { get { return partitioningType; } }
+        /// <summary>
+        /// The size of the disk in MB.
+        /// </summary>
+        public int Size { get; }
+        /// <summary>
+        /// List of partitions
+        /// </summary>
+        public List<ManagedPartition> Partitions { get { return parts; } }
```
Comment (quajak) [inline source/Cosmos.System2/FileSystem/Disk.cs:24]: Is the list necessary here? Cant we access the information from PartitioningType already? Also if we create a new Partition via the PartitioningType how do we update the list here?

File: source/Cosmos.System2/FileSystem/Disk.cs [hunk 15]
```diff
@@ -0,0 +1,152 @@
+﻿//#define COSMOSDEBUG
+
+using System;
+using System.Collections.Generic;
+using Cosmos.HAL.BlockDevice;
+using Cosmos.System.FileSystem.VFS;
+
+namespace Cosmos.System.FileSystem
+{
+    public class Disk
```
Comment (quajak) [inline source/Cosmos.System2/FileSystem/Disk.cs:10]: Can we merge `PartitioningType` and `Disk` into the same class, since they seem to overlap a lot.

**Review by quajak — CHANGES_REQUESTED** [review]

File: source/Cosmos.System2/FileSystem/Disk.cs (hunk 15 above)
Comment (MishaTY) [reply source/Cosmos.System2/FileSystem/Disk.cs:10]: We can, but the code will be messy if we do that.

Comment (quajak) [reply source/Cosmos.System2/FileSystem/Disk.cs:10]: How so?

**Pull request closed without merging**